Add option to use suda automatically if needed

doronbehar opened the pull request:
Fixes #7

EDIT:

This is kind of ugly but it works pretty good for me. Let me know what do you think, especially on those `cabbrev`iations.

doronbehar committed:
Add option to use suda automatically if needed

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,31 @@ function! suda#read(expr, ...) abort range
+function! suda#write_wrapper(action)
+  try 
+    execute('w' . a:action)
+  catch /^Vim\%((\a\+)\)\=:E45/
```
**must**

This regex will hit on `E450`, `E451` or whatever. It must be fit exactly on `E45` I guess.

doronbehar replied:
Agree, I couldn't find a way to do that, I tried adding `:` and `$` at the end of the pattern but it wasn't matched. I've taken that straight from the docs (`:help catch`). Are you sure it doesn't match only this specific error?

lambdalisue replied:
> Agree, I couldn't find a way to do that, I tried adding : and $ at the end of the pattern but it wasn't matched.

I didn't get what you mean by "wasn't matched".

The entire error message in my environment was 

```
Vim(write):E45: 'readonly' option is set (add ! to override)
```

and I think adding `:` works on my environment.

<img width="535" alt="1  tmux 2019-05-06 14-24-29" src="https://user-images.githubusercontent.com/546312/57207813-d9594180-700a-11e9-8af6-952ba011a756.png">

Btw, the error message was confirmed with

```vim
function! Test() abort
  try
    execute('w /some/file/which/is/not/writable')
  catch
    echomsg v:exception
  endtry
endfunction
```

> I've taken that straight from the docs

Yeah, I knew that the Vim's help says it in a bad regex 😢

doronbehar replied:
Right, `:` works. Don't know why I was under the impression it doesn't.

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,31 @@ function! suda#read(expr, ...) abort range
   endtry
 endfunction
 
+function! suda#write_wrapper(action)
```
**must**

Add `abort` on **any** functions.

**This rule must be applied to other functions as well** (I'm not going to review all for this).

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,31 @@ function! suda#read(expr, ...) abort range
+    let l:pat = '^' . g:suda#prefix
+  else
+    let l:pat = '^suda://*'
+  endif
```
may

You can write above block for a single simple line 👍 

```vim
let pat = '^' . get(g:, 'suda#prefix', 'suda://*')
```

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,31 @@ function! suda#read(expr, ...) abort range
+endfunction
+
+function suda#smart_read()
+  let l:fpath = expand('%')
```
*should*

I don't use `l:` prefix *on this project* so please remove it.
Btw I recently change mind so I agree to use `l:` for a new project.

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,31 @@ function! suda#read(expr, ...) abort range
+  if !filereadable(l:fpath)
+    let l:tobedeleted = bufnr('%')
+    execute 'edit suda://%'
+    exe "bd! " . l:tobedeleted
```
**must**

Do NOT use a short one. Use `execute` and `bdelete` instead.

lambdalisue commented on plugin/suda.vim:
```diff
@@ -6,3 +6,21 @@ let g:loaded_suda = 1
+  command! W execute suda#write_wrapper('')
+  command! Wa execute suda#write_wrapper('a')
+  command! Wq execute suda#write_wrapper('q')
+  command! Waq execute suda#write_wrapper('aq')
```
*should*

Do NOT use short names. Use descriptive one like `SudaWrite`, `SudaWriteAll`, `SudaWriteQuit`, `SudaWriteAllQuit` if there are not enough reasons to use the short one.

doronbehar replied:
The idea is, that with `g:suda_hijack_write` set, when a user types `:w` (lowercase) it will be replaced by the corresponding abbreviation to the command and he won't even feel that he's using the almighty suda plugin.

lambdalisue replied:
> he won't even feel that he's using the almighty suda plugin.

Even user won't use those commands, those commands will be popped up on the command completions and `W`, `Wa`, ...etc don't tell which plugin uses that command. If the command names are `SudaWrite` or whatever, users will notice that the commands are defined from suda.vim

doronbehar replied:
Fine with me :cat2:

lambdalisue commented on plugin/suda.vim:
```diff
@@ -6,3 +6,21 @@ let g:loaded_suda = 1
+if exists('g:suda_smart_read')
+  augroup suda_smart_read
+    au!
+    au BufReadCmd * nested call suda#smart_read()
```
**must**

Do NOT use short name. Use `autocmd` instead.

lambdalisue commented on plugin/suda.vim:
```diff
@@ -6,3 +6,21 @@ let g:loaded_suda = 1
+  command! W execute suda#write_wrapper('')
+  command! Wa execute suda#write_wrapper('a')
+  command! Wq execute suda#write_wrapper('q')
+  command! Waq execute suda#write_wrapper('aq')
```
**must**

You don' need `execute` here. Eg.

```vim
command! W suda#write_wrapper('')
```

doronbehar replied:
I think I am, I tried removing it and I got:

```
E492: Not an editor command: suda#write_wrapper('')
```

lambdalisue replied:
My bad. use `call` like

```vim
command! W call suda#write_wrapper('')
```

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,31 @@ function! suda#read(expr, ...) abort range
+  catch /^Vim\%((\a\+)\)\=:E45/
+    execute('w' . a:action . ' suda://%')
+  endtry
+endfunction
```
may

I feel `action` is a bit complicated because it must be trailing text for existing commands.

How about using the following more generic function instead

```vim
function! suda#execute_with_fallback(expr, fallback) abort
  try
    execute(a:expr)
  catch /^Vim\%((\a\+)\)\=:E45:/
    execute(a:fallback)
  endtry
endfunction

" ---

command! W suda#execute_with_fallback('w %', 'w suda://%')
```

doronbehar replied:
I've updated it to something a little different I hope you'll like as well, along with only `w` and `wq` abbreviations - no `wa` or `wqa` since I don't we think we should handle that as well.

lambdalisue reviewed (changes requested):
Thank you for the PR 👍 Reviewed with the following rules

- **must** - You must fix
- *should* - You should fix or explain why the code should be look like
- may - Just suggestion

doronbehar committed:
Add abort to all new functions

doronbehar committed:
Don't use short names

doronbehar committed:
Don't use l: prefix for variables

doronbehar committed:
Replace a:action with a:write_cmd in write_wrapper

Don't abbreviate `wa` or `wqa` since we shouldn't handle multiple files
needed to be written with sudo.

doronbehar commented:
One more thing: I noticed the `smart_read` is not smart enough, it doesn't exclude directories for example. It handles truely unreadable files fine but from some reason, it makes normal, readable files appear as empty. I'm working on that now.

doronbehar committed:
smart_read: handle directories and readable file

Use a b:suda_already_read variable to stop the function from being
called recursively.

doronbehar committed:
Replace BufReadCmd with BufEnter for smart_read

This way we don't have to read normal files for ourselves in
suda#smart_read

doronbehar commented:
I've fixed the issue mentioned in my previous comment. One last thing I'm not totally sure of is about the new settings I've added in `plugin/suda.vim`:

I noticed that if I put for exmaple `let g:suda_smart_read=1`  in a file inside my `&runtimepath`'s `plugin/` directory, when `plugin/suda.vim` is loaded, exists(`g:suda_smart_read`) is false. But If I put my `let g:suda_smart_read=1` in my `init.vim` it works fine. Perhaps we should use `after/plugin/`? BTW the same thing goes for `suda#prefix` as well.

Further more, maybe I should use `get(g:, ` instead? This will be more consistent with your way of reading settings..

doronbehar committed:
smart_read: make it aware of several edge cases

Pass to the main function the <abuf> and <afile> arguments so they won't
be retrieved in it.
Use these variables to keep track after already checked buffers using the
dictionary `s:smart_read_checked` - we don't even continue if already
checked buffer is encountered.

doronbehar committed:
Handle writeable nonexisting files

doronbehar committed:
Filter files with any URI pattern

Example: fugitive:// - He doesn't set buftype to anything although I
think he should.

lambdalisue commented on plugin/suda.vim:
```diff
@@ -6,3 +6,17 @@ let g:loaded_suda = 1
+if exists('g:suda_smart_read')
+  augroup suda_smart_read
+    autocmd!
+    autocmd BufEnter * nested call suda#smart_read(expand('<abuf>'), expand('<afile>'))
```
**must**

Do NOT use `BufEnter` for this purpose. It is called too often for this purpose.
I think `BufReadPre` may with `BufNewFile` is best.

lambdalisue commented on autoload/suda.vim:
```diff
@@ -93,6 +93,33 @@ function! suda#read(expr, ...) abort range
+    execute "bdelete! " . a:bufn
+  endif
+endfunction
+
```
**must**

This function does

1. Skip if the buffer has already checked
2. Skip if the buffer name starts from scheme (e.g. `fugitive://`)
3. Skip if the buffer is not a file
4. Skip if the buffer is new one
5. Skip if the file of the buffer does not exist but writable
6. Use `suda://` prefix instead if the buffer is not readable
7. Skip if the buffer is readable

But using `BufReadPre` will solves 1-5 I guess while

1. `BufReadPre` is not used if the file doesn' exist
2. `BufReadPre` is *usually* not used with `BufReadCmd` which is used for scheme loading (e.g. `fugitive://`)

Additionaly, I think using `execute 'keepjumps keepalt file suda://%'` is enough instead for executing `edit/bdelete!` sequentially (not confirmed).

Anyway, using `BufReadPre` would change this function completely I guess so I'll review new one.

doronbehar replied:
You have no idea how hard it was for me to find the right event.. I wished `BufReadPre` could be used but it's not and I had my head wrapped around it for a whole day why until I've found this stack-exchange QA that gives a good enough explanation / workaround: https://vi.stackexchange.com/questions/13178/autocmd-bufadd-bufcreate-bufnew-arent-triggered-for-new-unnamed-buffers-ene

lambdalisue replied:
Now I tried and I got what you meant 👍 

# tl;dr

1. Remove unnecessary autocmd instead of caching the detection result per buffers
2. Use `is#` or `isnot#` to compare numbers
3. Check if the file is not readable or does not exist by `getftype()`
4. Use `keepalt keepjumps file` to replace the buffer name instead

In short, apply the following diff.

```diff
diff --git a/autoload/suda.vim b/autoload/suda.vim
index 8487e31..e444a36 100644
--- a/autoload/suda.vim
+++ b/autoload/suda.vim
@@ -94,32 +94,13 @@ function! suda#read(expr, ...) abort range
 endfunction
 
 function! suda#write_wrapper(write_cmd) abort
-  try 
+  try
     execute(a:write_cmd)
   catch /^Vim\%((\a\+)\)\=:E45:/
     execute(a:write_cmd . ' suda://%')
   endtry
 endfunction
 
-let s:smart_read_checked = {}
-
-function! suda#smart_read(bufn, fpath) abort
-  if get(s:smart_read_checked, a:bufn, 0)
-    let s:smart_read_checked[a:bufn] = 1
-    return
-  end
-  if !match(a:fpath, '^[a-z]\+://*') || &buftype != '' || a:fpath ==# ''
-    return
-  endif
-  if !filereadable(a:fpath)
-    if filewritable(fnamemodify(a:fpath, ':p:h')) ==# 2
-      return
-    endif
-    execute 'edit suda://%'
-    execute "bdelete! " . a:bufn
-  endif
-endfunction
-
 function! suda#write(expr, ...) abort range
   let path = s:expand_expression(a:expr)
   let options = extend({
@@ -161,7 +142,6 @@ function! suda#write(expr, ...) abort range
   endtry
 endfunction
 
-
 function! suda#BufReadCmd() abort
   call s:doautocmd('BufReadPre')
   let ul = &undolevels
@@ -226,6 +206,25 @@ function! suda#FileWriteCmd() abort
   endtry
 endfunction
 
+function! suda#BufEnter() abort
+  augroup suda_smart_read_local
+    autocmd! * <buffer=abuf>
+  augroup END
+  let bufname = expand('<afile>')
+  if !empty(&buftype) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
+    " Non file buffer
+    return
+  elseif filereadable(bufname)
+    " File is readable
+    return
+  elseif empty(getftype(bufname)) && filewritable(fnamemodify(bufname, ':p:h')) is# 2
+    " File does not exist and the directory is writable
+    return
+  endif
+  execute printf('keepalt keepjumps file %s%%', g:suda#prefix)
+  edit
+endfunction
+
 function! s:escape_patterns(expr) abort
   return escape(a:expr, '^$~.*[]\')
 endfunction
diff --git a/plugin/suda.vim b/plugin/suda.vim
index 71aec8f..2276acd 100644
--- a/plugin/suda.vim
+++ b/plugin/suda.vim
@@ -15,8 +15,20 @@ if get(g:, 'suda_hijack_write')
 endif
 
 if get(g:, 'suda_smart_read')
-  augroup suda_smart_read
+  augroup suda_smart_read_primary
     autocmd!
-    autocmd BufEnter * nested call suda#smart_read(expand('<abuf>'), expand('<afile>'))
-  augroup end
+    autocmd BufEnter * nested call suda#BufEnter()
+    autocmd VimEnter *
+          \ augroup suda_smart_read_primary |
+          \   autocmd! * |
+          \ augroup END
+  augroup END
+
+  augroup suda_smart_read_secondary
+    autocmd! * <buffer>
+    autocmd BufNew *
+         \ augroup suda_smart_read_local |
+         \   autocmd BufEnter <buffer=abuf> nested call suda#BufEnter() |
+         \ augroup END
+  augroup END
 endif
```

**NOTE** that I haven't try above with all situations you might consider so please test by yourself :+1:

# Experiment

## Environment

I made readable and unreadable files by

```
$ echo "READABLE" > readable
$ echo "UNREADABLE" > unreadable
$ sudo chown root readable unreadable
$ sudo chmod 600 unreadable
```

Then add the following code to my `.vimrc` to investigate events

```vim
augroup autocmd_investigation
  autocmd!
  autocmd BufNew * echomsg "BufNew"
  autocmd BufAdd * echomsg "BufAdd"
  autocmd BufEnter * echomsg "BufEnter"

  autocmd BufNewFile * echomsg "BufNewFile"

  autocmd BufFilePre * echomsg "BufFilePre"
  autocmd BufFilePost * echomsg "BufFilePost"

  autocmd BufReadPre * echomsg "BufReadPre"
  autocmd BufReadPost * echomsg "BufReadPost"

  autocmd FileReadPre * echomsg "FileReadPre"
  autocmd FileReadPost * echomsg "FileReadPost"

  autocmd VimEnter * echomsg "VimEnter"
  autocmd WinEnter * echomsg "WinEnter"
augroup end
```

# Result

## Read readable/unreadable on startup

### Vim
```
$ vim readable
BufReadPre
BufReadPost
BufEnter
VimEnter

$ vim unreadable
BufEnter
VimEnter
```

### Neovim
```
$ nvim readable
BufEnter
VimEnter

$ nvim unreadable
BufEnter
VimEnter
```

## Read readable/unreadable with `:e` command

### Vim

```vim
:e readable
BufNew
BufAdd
BufReadPre
BufReadPost
BufEnter

:e unreadable
BufNew
BufAdd
BufEnter
```

### Neovim

```vim
:e readable
BufNew
BufAdd
BufEnter

:e unreadable
BufNew
BufAdd
BufEnter
```

# Conclusion

It seems there are two distinct pathway. So my conclusion is

1. Use `BufEnter` to detect if argument files are readable or not before `VimEnter` event
2. Use `BufEnter` which is sequentially after `BufNew` event

doronbehar replied:
Wow, well done on this investigation. It seems to be functioning well with my initial testings. I think it'll be good to keep this PR open for maybe even a few weeks if you don't mind so I can test it thoroughly with my normal usage with Vim and push small updates as I go.

Now I'd like please just to make sure I understand what you've done in `plugin/suda.vim`:

Are you running on `BufEnter` `suda#BufEnter()` and then remove this `autocmd` (addressing it through the `augroup` named `suda_smart_read_primary`)? If my understanding is correct, that's f*cking creative! Way better then those cheap workarounds in the [stack-exchange thread I mentioned](https://vi.stackexchange.com/questions/13178/autocmd-bufadd-bufcreate-bufnew-arent-triggered-for-new-unnamed-buffers-ene).

If my 1st assumption is correct, then I'd assume you've done a similar thing in `augroup suda_smart_read_secondary` but for `:e` commands and therfor, without `VimEnter` which is not fired. However, I don't understand: Why did you create another `augroup suda_smart_read_local` inside it? Inspired by what you've done in `augroup suda_smart_read_primary` I would do the same for when `:e` is used but using these events:

- Instead of `BufEnter`: `BufNew`
- Instead of `VimEnter`: `BufAdd`

lambdalisue replied:
>  I think it'll be good to keep this PR open for maybe even a few weeks if you don't mind so I can test it thoroughly with my normal usage with Vim and push small updates as I go.

Totally agree with your suggestion 👍 There might be some edge cases.

> Are you running on BufEnter suda#BufEnter() and then remove this autocmd (addressing it through the augroup named suda_smart_read_primary)? If my understanding is correct, that's f*cking creative! Way better then those cheap workarounds in the stack-exchange thread I mentioned.

Yes. This is a kind hacky way but quite effective. Note that Vim 8.1.1113 or Neovim 0.4.0-dev has `autocmd-once` so this kind of hack will not be required in the future.

> If my 1st assumption is correct, then I'd assume you've done a similar thing in augroup suda_smart_read_secondary but for :e commands and therfor, without VimEnter which is not fired. However, I don't understand: Why did you create another augroup suda_smart_read_local inside it? Inspired by what you've done in augroup suda_smart_read_primary I would do the same for when :e is used but using these events:

Ah, it seems your sense is incredible. Now I think this *once-like hack* is not required for the second one. Using `BufAdd` without using `BufNew` is enough I guess. Anyway, the above code is not effective so please optimize it 👍

doronbehar replied:
> Ah, it seems your sense is incredible. Now I think this once-like hack is not required for the second one. Using `BufAdd` without using `BufNew` is enough I guess. Anyway, the above code is not effective so please optimize it :+1:

Thanks :smile:. I must say though, that this hack is so slippery! I wonder how it took so long to feel the need for `autocmd-once`.. Anyway, I tried to rewrite what you did in somewhat more symmetric way for both `:e` and `$EDITOR file` situations:

```vim
if get(g:, 'suda_smart_read')
  augroup suda_smart_read_primary
    autocmd!
    autocmd BufEnter * nested call suda#BufEnter(expand('<afile>'))
    autocmd VimEnter *
          \ augroup suda_smart_read_primary |
          \   autocmd! * |
          \ augroup END
  augroup END

  augroup suda_smart_read_secondary
    autocmd!
    autocmd BufNew * nested call suda#BufEnter(expand('<afile>'))
    autocmd BufAdd *
          \ augroup suda_smart_read_secondary |
          \   autocmd! * |
          \ augroup END
  augroup END
endif
```

But, the problem is, that since we have to use `nested`, and `suda#BufEnter()` calls `edit`, I always get double `suda://` prefixes: `suda://suda://%`. Somehow our `match()` check misses this and there's nothing I could do to mitigate that.. I understand now why you have used an additional group `suda_smart_read_local`.

I think I'll revert this to the way it was in your original suggestion, unless you have in mind a way to optimize my approach in a way that would actually work. Or, we can revert it to the way it was with the dictionary cache of buffer numbers checked which I still think is not that bad considering it is probably the most readable / understandable for anyone who may encounter this in the future..

lambdalisue replied:
> But, the problem is, that since we have to use nested, and suda#BufEnter() calls edit, I always get double suda:// prefixes: suda://suda://%. Somehow our match() check misses this and there's nothing I could do to mitigate that.. I understand now why you have used an additional group suda_smart_read_local.

Probably `expand('<afile>')` is empty on `BufNew`. Try `BufAdd` instead.

Additionally, removing `suda_smart_read_secondary` is wrong while the second one should be applied to second, third, ... buffers as well (try `:e unreadable1` then `:e unreadable2` to see what I mean). So the correct one would be

```vim
if get(g:, 'suda_smart_read')
  augroup suda_smart_read_primary
    autocmd!
    autocmd BufEnter * nested call suda#BufEnter(expand('<afile>'))
    autocmd VimEnter *
          \ augroup suda_smart_read_primary |
          \   autocmd! * |
          \ augroup END
  augroup END

  augroup suda_smart_read_secondary
    autocmd!
    autocmd BufAdd * nested call suda#BufEnter(expand('<afile>'))
  augroup END
endif
```

But I'm not sure the following points

- Is `expand('<afile>')` returns value in `BufAdd` event?
    - May need `expand('<amatch>')`
- Is `file suda://% | edit` works as expect?

So above code might not work.

doronbehar replied:
That doesn't work just like my solution, unfortunately. I think this leaves us 2 options. Here they are summarized, along with the pros and cons IMO:

1. Use your original solution with both `augroup suda_smart_read_primary` and `augroup suda_smart_read_secondary` with `augroup suda_smart_read_local`.
   * Pros: Almost no unnecessary `autocmd`s (only `BufNew`).
   * Cons: Sort of complicated or in other words - very hard to understand for other developers when / if they'll read it.
2. Use the internal cache of buffers already checked.
   * Pros: Very simple to understand
   * Cons: ?

Anyway, Let me know what is your final opinion.

lambdalisue replied:
OK. Try the following.

```diff
diff --git a/autoload/suda.vim b/autoload/suda.vim
index 8487e31..069142f 100644
--- a/autoload/suda.vim
+++ b/autoload/suda.vim
@@ -94,32 +94,13 @@ function! suda#read(expr, ...) abort range
 endfunction
 
 function! suda#write_wrapper(write_cmd) abort
-  try 
+  try
     execute(a:write_cmd)
   catch /^Vim\%((\a\+)\)\=:E45:/
     execute(a:write_cmd . ' suda://%')
   endtry
 endfunction
 
-let s:smart_read_checked = {}
-
-function! suda#smart_read(bufn, fpath) abort
-  if get(s:smart_read_checked, a:bufn, 0)
-    let s:smart_read_checked[a:bufn] = 1
-    return
-  end
-  if !match(a:fpath, '^[a-z]\+://*') || &buftype != '' || a:fpath ==# ''
-    return
-  endif
-  if !filereadable(a:fpath)
-    if filewritable(fnamemodify(a:fpath, ':p:h')) ==# 2
-      return
-    endif
-    execute 'edit suda://%'
-    execute "bdelete! " . a:bufn
-  endif
-endfunction
-
 function! suda#write(expr, ...) abort range
   let path = s:expand_expression(a:expr)
   let options = extend({
@@ -161,7 +142,6 @@ function! suda#write(expr, ...) abort range
   endtry
 endfunction
 
-
 function! suda#BufReadCmd() abort
   call s:doautocmd('BufReadPre')
   let ul = &undolevels
@@ -226,6 +206,27 @@ function! suda#FileWriteCmd() abort
   endtry
 endfunction
 
+function! suda#BufEnter() abort
+  let bufname = expand('<amatch>')
+  if !empty(&buftype) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
+    " Non file buffer
+    return
+  elseif filereadable(bufname)
+    " File is readable
+    return
+  elseif empty(getftype(bufname)) && filewritable(fnamemodify(bufname, ':p:h')) is# 2
+    " File does not exist and the directory is writable
+    return
+  endif
+  let bufnr = bufnr('%')
+  execute printf(
+        \ 'keepalt keepjumps edit %s%s',
+        \ g:suda#prefix,
+        \ fnamemodify(bufname, ':p'),
+        \)
+  execute printf('silent! %dbwipeout', bufnr)
+endfunction
+
 function! s:escape_patterns(expr) abort
   return escape(a:expr, '^$~.*[]\')
 endfunction
diff --git a/plugin/suda.vim b/plugin/suda.vim
index 71aec8f..36d4f4c 100644
--- a/plugin/suda.vim
+++ b/plugin/suda.vim
@@ -15,8 +15,20 @@ if get(g:, 'suda_hijack_write')
 endif
 
 if get(g:, 'suda_smart_read')
-  augroup suda_smart_read
+  augroup suda_smart_read_primary
     autocmd!
-    autocmd BufEnter * nested call suda#smart_read(expand('<abuf>'), expand('<afile>'))
-  augroup end
+    autocmd BufEnter * nested call suda#BufEnter()
+    autocmd VimEnter *
+          \ augroup suda_smart_read_primary |
+          \   autocmd! * |
+          \ augroup END
+  augroup END
+
+  augroup suda_smart_read_secondary
+    autocmd! * <buffer>
+    autocmd BufNew *
+        \ augroup suda_smart_read_local |
+        \   autocmd BufEnter <buffer=abuf> nested call suda#BufEnter() |
+        \ augroup END
+  augroup END
 endif
```

I changed several points from the previous one because of

- `BufEnter` after `BufNew` hack is required because `&buftype` on `BufNew/BufAdd` refers to the previous buffer one
- `edit` then `bwipeout` hack (like what you did in previous one!) is required because `file` does not work on existing buffer
- `suda_smart_read_local` group must NOT be removed otherwise `:e unreadable | e brabra | e unreadable` does not work

lambdalisue commented on plugin/suda.vim:
```diff
@@ -6,3 +6,17 @@ let g:loaded_suda = 1
   call suda#init()
 endif
+
+if exists('g:suda_hijack_write')
```
may

Using `get` is better because users can control on/off by the value like

```vim
if get(g:, 'suda_hijack_write')
    " ...
endif
```

```vim
let g:suda_hijack_write = 0  " Disable
" let g:suda_hijack_write = 1  " Enable
```

lambdalisue reviewed (changes requested):
I'm not really sure but using `BufReadPre` would make things really simple so please reconsider 👍 

> I noticed that if I put for exmaple let g:suda_smart_read=1 in a file inside my &runtimepath's plugin/ directory, when plugin/suda.vim is loaded, exists(g:suda_smart_read) is false. But If I put my let g:suda_smart_read=1 in my init.vim it works fine. Perhaps we should use after/plugin/? BTW the same thing goes for suda#prefix as well.

You should NOT rely on the sourcing order of files in `plugin/` directory. That's why `init.vim` is called prior to any files in `plugin/` and used for initial settings. Using `after/plugin` is quite a bad idea for providing initial values because users cannot override the settings.

> Further more, maybe I should use get(g:, instead? This will be more consistent with your way of reading settings..

Yes for settings which user may touch. See my comment.

doronbehar committed:
Match exact readonly error

doronbehar committed:
Use call instead of execute in abbreviations

doronbehar committed:
Use  and not exists for settings

doronbehar committed:
Use better names for new commands

doronbehar committed:
Apply Patch from PR, see discussion

doronbehar committed:
Use autocmds as suggested in PR review

doronbehar committed:
Handle another edge case of nonexisting files

By recursively reading parent directories, check if we can write to
any of the parent directories so when a new buffer is created, e.g.
`non/existing/dir/file.txt`, if our current working directory is
writeable, we return as well.

doronbehar commented:
Thanks for your help in making this feature a reality @lambdalisue. I'll mark the review above as resolved and push the changes accordingly.

Besides our discussed dilemmas above, I have one more UX issue I'd like to seek advice with you:

Besides this plugin, I have the plugin [eunuch](https://github.com/tpope/vim-eunuch) installed as well. I use it mostly for `:Mkdir`, `:Rename` and `:Move`. After experimenting with our new feature, I noticed that on there is a sort of collision with a [command defined by eunuch (`:W` and `:Wall`)](https://github.com/tpope/vim-eunuch/blob/e066a0999e442d9d96f24ad9d203b1bd030ef72e/doc/eunuch.txt#L88). The problem is that from some reason I end up in Insert mode after every write. Additionally, the writes seem to be slow because they are wrapped in a function.

Anyway, I was thinking that maybe all of this write wrapper is not that necessary and maybe we can just make this whole feature a 1 setting thing and we'll replace buffers with their `suda://`s even if they are not writeable. In code, that means:

<details><summary>See this patch:</summary>

```patch
diff --git i/autoload/suda.vim w/autoload/suda.vim
index 350a44c..cc4765a 100644
--- i/autoload/suda.vim
+++ w/autoload/suda.vim
@@ -93,14 +93,6 @@ function! suda#read(expr, ...) abort range
   endtry
 endfunction
 
-function! suda#write_wrapper(write_cmd) abort
-  try
-    execute(a:write_cmd)
-  catch /^Vim\%((\a\+)\)\=:E45:/
-    execute(a:write_cmd . ' suda://%')
-  endtry
-endfunction
-
 function! suda#write(expr, ...) abort range
   let path = s:expand_expression(a:expr)
   let options = extend({
@@ -212,8 +204,8 @@ function! suda#BufEnter() abort
   if !empty(&buftype) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
     " Non file buffer
     return
-  elseif filereadable(bufname)
-    " File is readable
+  elseif filereadable(bufname) && filewritable(bufname)
+    " File is readable and writeable
     return
   elseif empty(getftype(bufname))
     " if file doesn't exist, we search for a all directories up it's path to
diff --git i/plugin/suda.vim w/plugin/suda.vim
index 2276acd..2e926fa 100644
--- i/plugin/suda.vim
+++ w/plugin/suda.vim
@@ -7,13 +7,6 @@ if get(g:, 'suda_startup', 1)
   call suda#init()
 endif
 
-if get(g:, 'suda_hijack_write')
-  cabbrev w SudaW
-  cabbrev wq SudaWq
-  command! SudaW call suda#write_wrapper('w')
-  command! SudaWq call suda#write_wrapper('wq')
-endif
-
 if get(g:, 'suda_smart_read')
   augroup suda_smart_read_primary
     autocmd!
```
</details>

lambdalisue commented:
> After experimenting with our new feature, I noticed that on there is a sort of collision with a command defined by eunuch (:W and :Wall).

That's why I asked to add `Suda` prefix to our command. That's why I beleive that all plugins should NOT define [this kind of alias](https://github.com/tpope/vim-eunuch/blob/master/plugin/eunuch.vim#L297-L299). My advice for this is "Do NOT use tpope's plugins or do NOT use others' plugins 😇"

> Anyway, I was thinking that maybe all of this write wrapper is not that necessary and maybe we can just make this whole feature a 1 setting thing and we'll replace buffers with their suda://s even if they are not writeable. In code, that means:

So suda will automatically use "suda://" prefix when

1. The file is not readable
2. The file is not writable

right? It sounds reasonable 👍

doronbehar committed:
Remove suda_hijack_write and check if filewriteble

This is much better then the try catch thing, see this comment for
reference:
https://github.com/lambdalisue/suda.vim/pull/13#issuecomment-491703332

doronbehar commented:
Oh right then, I think this PR is fully ready. I'm pushing the patch I've showed [in my previous comment](https://github.com/lambdalisue/suda.vim/pull/13#issuecomment-491703332). There should be added some docs as well before this can be merged but I'm postponing this in the meantime because I might change my mind again as for some aspects of the behaviour as I'll be testing this in the next few following days..

lambdalisue commented on autoload/suda.vim:
```diff
@@ -199,6 +198,41 @@ function! suda#FileWriteCmd() abort
+      endif
+      " we keep on the iteration by going up the tree
+      let parent = fnamemodify(parent, ':h')
+    endwhile
```
:question:

I don't get why the recursive check is required here.

1. The file exists
    1. The file is readable and writable then we do not need `suda://`
    2. Otherwise we need `suda://`
2. The file does not exist
    1. The parent is not writable then we need `suda://`
    2. The grandparent is writable but the parent is not writable, we **still** need `suda://`
    3. The parent is writable but not readable, we **won't** need `suda://`

I think `empty(getftype(bufname)) && filewritable(bufname)` (or `filewritable(fnamemodify(bufname, ':p:h'))` maybe?) is enough.

doronbehar replied:
Because sometimes, you may find yourself opening a file under a writeable directory e.g.  `~/all/of/these/directories/do/not/exist.txt` and then you would like to edit that buffer (of course without `suda://`) and when it is ready for save, `:!mkdir -p %:p:h` and then `:w`.

lambdalisue replied:
Even without suda.vim, `:e ~/all/of/these/directories/do/not/exist.txt` does not work anyway so suda.vim should NOT consider that case I think.

doronbehar replied:
> `:e ~/all/of/these/directories/do/not/exist.txt` does not work anyway

What do you mean by 'does not work anyway'?

lambdalisue replied:
You've already explained it by yourself. Simple `:w` won't work.

I believe that procedure without automatic `suda://` prefix is more simple

### Unwritable without automatic `suda://` prefix

```
:e /all/of/these/directories/do/not/exist.txt
:!sudo mkdir -p %:p:h
password
:w suda://%
```

### Unwritable with automatic `suda://` prefix

```
:e /all/of/these/directories/do/not/exist.txt
:!sudo mkdir -p <C-r>=fnamemodify(matchstr(expand('%'), '^suda://\zs.*'), ':p:h')
password
:w
```

lambdalisue replied:
My bad. I think I got the reason for this.

Well, could you use the following instead in case? It made sure that the loop stop on root.

```vim
    " if file doesn't exist, we search for a all directories up it's path to
    " see if each one of them is writable, if not, we `return`
    let parent = fnamemodify(bufname, ':p')
    while parent !=# fnamemodify(parent, ':h')  " stop loop on root
      let parent = fnamemodify(parent, ':h')
      if filewritable(parent) is# 2
        " reached a writeable directory - no need to switch to suda://
        return
      elseif !filereadable(parent) && isdirectory(parent)
        " reached a readable parent directory which is not writeable so it
        " means we finally can be sure we need to replace our buffer with a
        " suda:// .
        break
      endif
    endwhile
```

lambdalisue reviewed (commented):
Oops. I reviewed but have forgotten to submit 😨 Sorry for late.

doronbehar commented:
There's one more edge case I'd like to discuss with you:

I noticed that with our current set of `autocmd`s, if I open multiple unwriteable files with the new feature enabled, only one of them gets the `suda://` prefix. This brings me back to our original discussion were I tend to think the `BufNew` approach, with an `s:` variable for caching checked buffers is better. For me, this method feels much more reassuring in it's technical design.. What do you think?

lambdalisue commented:
> This brings me back to our original discussion were I tend to think the BufNew approach, with an s: variable for caching checked buffers is better. For me, this method feels much more reassuring in it's technical design.. What do you think?

Well, I believe that using `s:` is not the best design for this case. It seems simple but actually is complex. So I would love to try really hard to find a way without using `s:` approach.

Could you write step by step procedure for your case?

doronbehar commented:
> Could you write step by step procedure for your case?

Just open multiple unwriteable files at once from the command line, e.g. `nvim /etc/profile /etc/environment`.

lambdalisue commented:
> Just open multiple unwriteable files at once from the command line, e.g. nvim /etc/profile /etc/environment.

I got.

```
$ nvim ~/wk/unreadable ~/wk/unreadable2
BufEnter
VimEnter
:next 
BufEnter
```

I'll think about this but now I feel cache style is required to handle this case with "user-friendly" interface...

doronbehar committed:
Make sure ':h' loop stops on root

doronbehar committed:
Revert "Apply Patch from PR, see discussion"

This reverts commit 422ef745318968477bdfa74fd4ea7d470400573d which
introduced a sophisticated but insufficient mechanism for automatically
replacing buffers with their suda:// equivalents using only `autocmd`s.
This commit brings back the old behaviour of using a dictionary as a
cache for inspected files so even when Vim is launched from the CLI with
multiple suda needed files, it will replace all them with suda://%.

doronbehar commented:
> I'll think about this but now I feel cache style is required to handle this case with "user-friendly" interface...

Please keep me updated. I'm changing the branch now to the way it was with only one `autocmd` and the dictionary cache based check, I'm not pushing it though. I'll keep you updated if I find multiple edge cases with that approach as well.

doronbehar committed:
Check if isdirectory for extra safety

In case no directory handler plugin is available, suda may attempt to
replace current buffer while it is actually a directory. The test
!empty(&buftype) may fail in such case. Perhaps it is so because we've
started using BufEnter

doronbehar committed:
Actually use the cache dict

Until now everything worked because our tests made the function return
anyway but now it works faster using the cache dictionary.

lambdalisue commented on plugin/suda.vim:
```diff
@@ -6,3 +6,10 @@ let g:loaded_suda = 1
+    autocmd!
+    autocmd BufEnter * nested call suda#BufEnter()
+  augroup end
+endif
```
may

Let's say it `suda_smart_edit` while it is to read and write.

doronbehar replied:
Agree.

lambdalisue commented on autoload/suda.vim:
```diff
@@ -199,6 +198,48 @@ function! suda#FileWriteCmd() abort
+        \ fnamemodify(bufname, ':p'),
+        \)
+  execute printf('silent! %dbwipeout', bufnr)
+endfunction
```
**must**

Please replace the function to the code below and confirm the behavior

```vim
function! suda#BufEnter() abort
  if exists('b:suda_smart_edit_checked')
    return
  endif
  let b:suda_smart_edit_checked = 1
  let bufname = expand('<afile>')
  if !empty(&buftype)
        \ || empty(bufname)
        \ || match(bufname, '^[a-z]\+://*') isnot# -1
        \ || isdirectory(bufname)
    " Non file buffer
    return
  elseif filereadable(bufname) && filewritable(bufname)
    " File is readable and writeable
    return
  elseif empty(getftype(bufname))
    " if file doesn't exist, we search for a all directories up it's path to
    " see if each one of them is writable, if not, we `return`
    let parent = fnamemodify(bufname, ':p')
    while parent !=# fnamemodify(parent, ':h')
      let parent = fnamemodify(parent, ':h')
      if filewritable(parent) is# 2
        return
      elseif !filereadable(parent) && isdirectory(parent)
        break
      endif
    endwhile
  endif
  let bufnr = str2nr(expand('<abuf>'))
  execute printf(
        \ 'keepalt keepjumps edit %s%s',
        \ g:suda#prefix,
        \ fnamemodify(bufname, ':p'),
        \)
  execute printf('silent! %dbwipeout', bufnr)
endfunction
```

Above fixed the followings

- Use `b:` instead of `s:` so that the variable is removed when the buffer has removed
  - Using `exists()` is enough because users won't touch `b:suda_smart_edit_checked`
- Call `isdirectory` at last because it would invoke OS's file system access
- Move `bufname` and `bufnr` initialization to the line just before requirements while `BufEnter` is invoked often and early return in Vim script is quite important
- Use `expand('<abuf>')` instead of `bufnr('%')` because we use `<afile>` instead of `%`
- Remove comments in a loop block because Vim parse these line repeatedly

doronbehar replied:
I agree with every word you've said there. According to my initial testings, everything is fine.

lambdalisue reviewed (changes requested):
OK. Let's use cache instead. But please use `b:` instead of `s:` so that the variable is automatically wiped out when the target buffer is wiped out or so on. Please check my comment and replace 👍

doronbehar committed:
Use buffer cache and optimize #BufEnter

doronbehar committed:
Rename augroup since it is for read and write

doronbehar committed:
Add to doc/suda.txt g:suda_smart_read

Make g:suda_startup a little bit clearer in relation to new setting
`g:suda_smart_read`.

doronbehar commented:
Oh and as for documentation, please see my last commit https://github.com/lambdalisue/suda.vim/pull/13/commits/5f0870a897393523c7f95a125820e7671f0b1452 .

lambdalisue commented on plugin/suda.vim:
```diff
@@ -8,7 +8,7 @@ if get(g:, 'suda_startup', 1)
 endif
 
 if get(g:, 'suda_smart_read')
```
may

Why don't you change the variable name as well?

doronbehar replied:
I forgot.

lambdalisue commented on doc/suda.txt:
```diff
@@ -105,13 +105,23 @@ VARIABLES					*suda-variables*
-	Call |suda#init()| automatically.
+	If set, call |suda#init()| automatically on buffers suda buffers,
+	according to prefix.
+	Default: 1
```
🎉

lambdalisue commented on doc/suda.txt:
```diff
@@ -105,13 +105,23 @@ VARIABLES					*suda-variables*
+	`suda://%` - depending on your |g:suda#prefix|, naturally. The check
+	is done only once for every buffer and it is designed to be optimized
+	as possible to you shouldn't feel any slowdown when openning buffers.
+	Default: 0
```
may

Variable name?

lambdalisue reviewed (approved):
Thanks for your wonderful works :tada: LGTM. But I wonder why you didn't change the variable name but augroup or so on. Are you sure to keep `g:suda_smart_read` instead of `g:suda_smart_edit`?

doronbehar committed:
Change name of main feature var

doronbehar committed:
doc: Small rephrasing fixes

lambdalisue commented:
Great work 🎉 Thanks to keeping the discussion 👍

doronbehar commented:
It was a pleasure :)
